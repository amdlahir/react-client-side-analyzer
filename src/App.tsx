import styles from './styles/app.module.css'
import { FileSelect } from './lib/components/fileSelect/fileSelect'
import csvWorker from './lib/workers/csvWorker.ts?worker'
import React, { useState } from 'react';
import { useCsvProcessor } from './lib/hooks/useCsvProcessor';
import { Statscard } from './Statscard';
import { CSVProcessor } from './lib/utils/csvProcessor';
import { ColumnStatsChart } from './lib/components/columnstatsChart/ColumnStatsChart';
import { Button, ButtonVariant } from './lib/components/button';
export type CsvWorker = typeof csvWorker;

function App() {
  const [files, setFiles] = useState<File[]>([]);
  const results = useCsvProcessor(files, csvWorker);
  const [selectedFile, setSelectedFile] = useState<CSVProcessor | undefined>(undefined);

  const isLoading = results.some(result => result.isLoading);
  const data = results
    .filter(result => result.isSuccess && result.data)
    .map(result => result.data);

  function handleFileChange(files: File[]) {
    setFiles(prevFiles => {
      const newFiles = files.filter(file => !prevFiles.some(existingFile => existingFile.name === file.name));
      return [...prevFiles, ...newFiles];
    });
  }

  function handleCloseChart() {
    setSelectedFile(undefined);
  }

  return (
    <React.Fragment>
      {selectedFile ?  (
        <React.Fragment>
          <div className={styles.screenOverlay} />
          <div className={styles.chartContainer} role="dialog">
            <ColumnStatsChart columnStats={selectedFile.stats} onClose={handleCloseChart} />
          </div>
        </React.Fragment>
      ) : null}
      <main className={styles.page}>
        <div className={styles.fileSelectContainer}>
          <FileSelect
            disabled={isLoading}
            onFilesSelected={handleFileChange}
            className={styles.fileSelect}>
            <FileSelect.Input
              accept=".csv,text/csv"
              multiple
            />
            <FileSelect.DropZone>
              {isLoading ? <p>Please wait while we process your files...</p> : (
                <p>
                  Add CSV files
                  <br />
                  <span>
                  (click or drag & drop files here)
                  </span>
                </p>
              )}
            </FileSelect.DropZone>
          </FileSelect>
          <section className={styles.statsContainer}>
            {data.map((fileData) => (
              fileData && (
                <Statscard 
                  key={fileData.processorId} 
                  data={fileData} 
                  onPrimaryBtnClick={() => setSelectedFile(fileData)}
                  onSecondaryBtnClick={() => setSelectedFile(fileData)}
                />
              )
            ))}
            {data && data.length > 0 ? (
              <div>
                <Button pill variant={ButtonVariant.SECONDARY} size="sm" text="Clear" onClick={() => setFiles([])} />
              </div>
            ) : null}
          </section>
        </div>
      </main>
    </React.Fragment>
  )
}

export default App
