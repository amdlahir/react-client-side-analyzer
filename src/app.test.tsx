import React from 'react';
import { describe, it, expect, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import { userEvent } from '@testing-library/user-event';
import App from './App';
import { mockResults } from './test/mocks/processedCsvData';

const mocks = vi.hoisted(() => {
  return {
    mockUseQueriesHook: vi.fn(),
  };
});

vi.mock('@tanstack/react-query', () => ({
  useQueries: mocks.mockUseQueriesHook,
}))

describe('Application', () => {
  it('should render the file selector on initial render', () => {
    mocks.mockUseQueriesHook.mockReturnValue([])
    render(<App />);
    expect(screen.getByText(/Add csv file/i)).toBeInTheDocument();
  })

  it('shoud render multiple csv file stats cards', async () => {
    mocks.mockUseQueriesHook.mockReturnValue(mockResults)
    render(<App />)
    for (const result of mockResults) {
      expect(screen.getByText(`File: ${result.data.file.name}`)).toBeInTheDocument();
    }
    const viewStatsButtons = screen.getAllByText(/view stats/i)
    expect(viewStatsButtons.length).toBe(mockResults.length);
    //expand card 1
    await userEvent.click(viewStatsButtons[0])
    Object.keys(mockResults[0].data.stats).forEach(colName => {
      expect(screen.getByText(colName)).toBeInTheDocument();
    })
    expect(screen.getAllByText('string').length).toBe(2);
    expect(screen.getAllByText('number').length).toBe(2);
    expect(screen.getAllByText('0').length).toBe(3);
    expect(screen.getAllByText('100').length).toBe(1);
    expect(screen.getAllByText('200').length).toBe(1);
    expect(screen.getAllByText('300').length).toBe(1);
    expect(screen.getAllByText('400').length).toBe(1);
    //minimize card 1
    await userEvent.click(screen.getByText(/minimize/i))
    Object.keys(mockResults[0].data.stats).forEach(colName => {
      expect(screen.queryByText(colName)).not.toBeInTheDocument();
    })

    //expand card 2
    await userEvent.click(viewStatsButtons[1])
    Object.keys(mockResults[1].data.stats).forEach(colName => {
      expect(screen.getByText(colName)).toBeInTheDocument();
    })
    expect(screen.getAllByText('string').length).toBe(1);
    expect(screen.getAllByText('number').length).toBe(1);
    expect(screen.getAllByText('0').length).toBe(2);
    expect(screen.getAllByText('10').length).toBe(1);
    expect(screen.getAllByText('20').length).toBe(1);
    //minimize card 2
    await userEvent.click(screen.getByText(/minimize/i))
    Object.keys(mockResults[1].data.stats).forEach(colName => {
      expect(screen.queryByText(colName)).not.toBeInTheDocument();
    })
  })

  it('shoud render chart', async () => {
    mocks.mockUseQueriesHook.mockReturnValue(mockResults);
    render(<App />);
    const viewStatsButtons = screen.getAllByText(/view stats/i);
    await userEvent.click(viewStatsButtons[0]);
    await userEvent.click(screen.getByText(/view chart/i));
    expect(screen.getByText(/mean and standard deviation/i)).toBeInTheDocument();
    expect(screen.getByRole('dialog')).toMatchSnapshot();
  })
})