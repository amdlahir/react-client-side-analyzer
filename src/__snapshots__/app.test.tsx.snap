// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Application > shoud render chart 1`] = `
<div
  class="_chartContainer_4551aa"
  role="dialog"
>
  <div
    class="_container_ecf76d"
  >
    <div
      class="_titleContainer_ecf76d"
    >
      <h2
        class="_title_ecf76d"
      >
        Mean and Standard Deviation
      </h2>
      <button
        class="_btn_abafc2 _btn_secondary_abafc2 _btn_sm_abafc2 _pill_abafc2 _flat_abafc2"
        style="height: 25px; line-height: 25px; min-width: 25px; font-size: 0.8rem; border-radius: 50%; padding: 0px;"
        type="button"
      >
        X
      </button>
    </div>
    <div
      class="recharts-wrapper"
      style="position: relative; cursor: default; width: 700px; height: 500px;"
    >
      <svg
        class="recharts-surface"
        height="500"
        style="width: 100%; height: 100%;"
        viewBox="0 0 700 500"
        width="700"
      >
        <title />
        <desc />
        <defs>
          <clippath
            id="recharts1-clip"
          >
            <rect
              height="445"
              width="590"
              x="80"
              y="20"
            />
          </clippath>
        </defs>
        <g
          class="recharts-cartesian-grid"
        >
          <g
            class="recharts-cartesian-grid-horizontal"
          >
            <line
              fill="none"
              height="445"
              stroke="#ccc"
              stroke-dasharray="3 3"
              width="590"
              x="80"
              x1="80"
              x2="670"
              y="20"
              y1="465"
              y2="465"
            />
            <line
              fill="none"
              height="445"
              stroke="#ccc"
              stroke-dasharray="3 3"
              width="590"
              x="80"
              x1="80"
              x2="670"
              y="20"
              y1="353.75"
              y2="353.75"
            />
            <line
              fill="none"
              height="445"
              stroke="#ccc"
              stroke-dasharray="3 3"
              width="590"
              x="80"
              x1="80"
              x2="670"
              y="20"
              y1="242.5"
              y2="242.5"
            />
            <line
              fill="none"
              height="445"
              stroke="#ccc"
              stroke-dasharray="3 3"
              width="590"
              x="80"
              x1="80"
              x2="670"
              y="20"
              y1="131.25"
              y2="131.25"
            />
            <line
              fill="none"
              height="445"
              stroke="#ccc"
              stroke-dasharray="3 3"
              width="590"
              x="80"
              x1="80"
              x2="670"
              y="20"
              y1="20"
              y2="20"
            />
          </g>
          <g
            class="recharts-cartesian-grid-vertical"
          >
            <line
              fill="none"
              height="445"
              stroke="#ccc"
              stroke-dasharray="3 3"
              width="590"
              x="80"
              x1="227.5"
              x2="227.5"
              y="20"
              y1="20"
              y2="465"
            />
            <line
              fill="none"
              height="445"
              stroke="#ccc"
              stroke-dasharray="3 3"
              width="590"
              x="80"
              x1="522.5"
              x2="522.5"
              y="20"
              y1="20"
              y2="465"
            />
            <line
              fill="none"
              height="445"
              stroke="#ccc"
              stroke-dasharray="3 3"
              width="590"
              x="80"
              x1="80"
              x2="80"
              y="20"
              y1="20"
              y2="465"
            />
            <line
              fill="none"
              height="445"
              stroke="#ccc"
              stroke-dasharray="3 3"
              width="590"
              x="80"
              x1="670"
              x2="670"
              y="20"
              y1="20"
              y2="465"
            />
          </g>
        </g>
        <g
          class="recharts-layer recharts-cartesian-axis recharts-xAxis xAxis"
        >
          <line
            class="recharts-cartesian-axis-line"
            fill="none"
            height="30"
            orientation="bottom"
            stroke="#666"
            width="590"
            x="80"
            x1="80"
            x2="670"
            y="465"
            y1="465"
            y2="465"
          />
          <g
            class="recharts-cartesian-axis-ticks"
          >
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="30"
                orientation="bottom"
                stroke="#666"
                width="590"
                x="80"
                x1="227.5"
                x2="227.5"
                y="465"
                y1="471"
                y2="465"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="30"
                orientation="bottom"
                stroke="none"
                text-anchor="middle"
                width="590"
                x="227.5"
                y="473"
              >
                <tspan
                  dy="0.71em"
                  x="227.5"
                >
                  col_3
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="30"
                orientation="bottom"
                stroke="#666"
                width="590"
                x="80"
                x1="522.5"
                x2="522.5"
                y="465"
                y1="471"
                y2="465"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="30"
                orientation="bottom"
                stroke="none"
                text-anchor="middle"
                width="590"
                x="522.5"
                y="473"
              >
                <tspan
                  dy="0.71em"
                  x="522.5"
                >
                  col_4
                </tspan>
              </text>
            </g>
          </g>
        </g>
        <g
          class="recharts-layer recharts-cartesian-axis recharts-yAxis yAxis"
        >
          <line
            class="recharts-cartesian-axis-line"
            fill="none"
            height="445"
            orientation="left"
            stroke="#666"
            width="60"
            x="20"
            x1="80"
            x2="80"
            y="20"
            y1="20"
            y2="465"
          />
          <g
            class="recharts-cartesian-axis-ticks"
          >
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="445"
                orientation="left"
                stroke="#666"
                width="60"
                x="20"
                x1="74"
                x2="80"
                y="20"
                y1="465"
                y2="465"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="445"
                orientation="left"
                stroke="none"
                text-anchor="end"
                width="60"
                x="72"
                y="465"
              >
                <tspan
                  dy="0.355em"
                  x="72"
                >
                  -250
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="445"
                orientation="left"
                stroke="#666"
                width="60"
                x="20"
                x1="74"
                x2="80"
                y="20"
                y1="353.75"
                y2="353.75"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="445"
                orientation="left"
                stroke="none"
                text-anchor="end"
                width="60"
                x="72"
                y="353.75"
              >
                <tspan
                  dy="0.355em"
                  x="72"
                >
                  0
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="445"
                orientation="left"
                stroke="#666"
                width="60"
                x="20"
                x1="74"
                x2="80"
                y="20"
                y1="242.5"
                y2="242.5"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="445"
                orientation="left"
                stroke="none"
                text-anchor="end"
                width="60"
                x="72"
                y="242.5"
              >
                <tspan
                  dy="0.355em"
                  x="72"
                >
                  250
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="445"
                orientation="left"
                stroke="#666"
                width="60"
                x="20"
                x1="74"
                x2="80"
                y="20"
                y1="131.25"
                y2="131.25"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="445"
                orientation="left"
                stroke="none"
                text-anchor="end"
                width="60"
                x="72"
                y="131.25"
              >
                <tspan
                  dy="0.355em"
                  x="72"
                >
                  500
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="445"
                orientation="left"
                stroke="#666"
                width="60"
                x="20"
                x1="74"
                x2="80"
                y="20"
                y1="20"
                y2="20"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="445"
                orientation="left"
                stroke="none"
                text-anchor="end"
                width="60"
                x="72"
                y="20"
              >
                <tspan
                  dy="0.355em"
                  x="72"
                >
                  750
                </tspan>
              </text>
            </g>
          </g>
        </g>
        <g
          class="recharts-layer recharts-bar"
        >
          <g
            class="recharts-layer recharts-bar-rectangles"
          >
            <g
              class="recharts-layer recharts-bar-rectangle"
            >
              <path
                class="recharts-rectangle"
                d="M 109.5,309.25 h 236 v 44.5 h -236 Z"
                fill="#40a0a0"
                height="44.5"
                name="col_3"
                radius="0"
                width="236"
                x="109.5"
                y="309.25"
              />
            </g>
            <g
              class="recharts-layer recharts-bar-rectangle"
            >
              <path
                class="recharts-rectangle"
                d="M 404.5,220.24999999999997 h 236 v 133.50000000000003 h -236 Z"
                fill="#40a0a0"
                height="133.50000000000003"
                name="col_4"
                radius="0"
                width="236"
                x="404.5"
                y="220.24999999999997"
              />
            </g>
          </g>
          <g
            class="recharts-layer"
          >
            <g
              class="recharts-layer recharts-errorBars"
            >
              <g
                class="recharts-layer recharts-errorBar"
                stroke="#2a7070"
                stroke-width="2"
              >
                <line
                  x1="223.5"
                  x2="231.5"
                  y1="220.24999999999997"
                  y2="220.24999999999997"
                />
                <line
                  x1="227.5"
                  x2="227.5"
                  y1="398.25"
                  y2="220.24999999999997"
                />
                <line
                  x1="223.5"
                  x2="231.5"
                  y1="398.25"
                  y2="398.25"
                />
              </g>
              <g
                class="recharts-layer recharts-errorBar"
                stroke="#2a7070"
                stroke-width="2"
              >
                <line
                  x1="518.5"
                  x2="526.5"
                  y1="42.25000000000002"
                  y2="42.25000000000002"
                />
                <line
                  x1="522.5"
                  x2="522.5"
                  y1="398.25"
                  y2="42.25000000000002"
                />
                <line
                  x1="518.5"
                  x2="526.5"
                  y1="398.25"
                  y2="398.25"
                />
              </g>
            </g>
          </g>
        </g>
      </svg>
      <div
        class="recharts-tooltip-wrapper"
        style="visibility: hidden; pointer-events: none; position: absolute; top: 0px; left: 0px;"
        tabindex="-1"
      />
    </div>
  </div>
</div>
`;
